src/FakeItEasy/Creation/DummyCreationSession.cs

```diff
@@ -0,0 +1,26 @@
+﻿namespace FakeItEasy.Creation
+{
+    using System;
+    using System.Collections.Generic;
+
+    internal class DummyCreationSession
+    {
+        private readonly HashSet<Type> typesCurrentlyBeingResolved = new HashSet<Type>();
+
+        public bool TryBeginToResolveType(Type typeOfDummy)
+        {
+            if (this.typesCurrentlyBeingResolved.Contains(typeOfDummy))
+            {
+                return false;
+            }
+
+            this.typesCurrentlyBeingResolved.Add(typeOfDummy);
+            return true;
+        }
+
+        public void OnSuccessfulResolve(Type typeOfDummy)
+        {
+            this.typesCurrentlyBeingResolved.Remove(typeOfDummy);
+        }
+    }
+}

```

Comment:
So, `DummyCreationSession` is just a wrapper around a hashset to keep track of which types are being resolved?

This means that the strategy cache will be shared between all dummy resolutions. I guess it makes sense, but can't it have unwanted side effects?